Color: Add 'colorSpace' argument for getters/setters

donmccurdy opened the pull request:
**Related**

- #23430 *(edited)*
- #22346 
- #22275
- #19169
- #11337

**Description**

Certain ways of expressing colors — hexadecimal and CSS strings — can safely be assumed to be in the sRGB color space. Hexadecimal syntax implies sRGB by convention, and CSS colors are _always_ sRGB except where otherwise specified in the (very recent) CSS Color Module Level 4.

However, colors passed to three.js shader must be Linear-sRGB. Because hexadecimal and CSS colors are not (by convention or specification) in the required color space, I believe we should take the step of converting them from sRGB to Linear-sRGB on the user's behalf. Importantly, this ensures that hexadecimal colors from a color picker, or CSS colors from a stylesheet, can be used directly in three.js APIs. They will appear correctly without manual conversion, assuming `.outputEncoding = sRGBEncoding` is also set.

The changes provided by this PR are disabled by default (`ColorManagement.legacyMode = true`), but can be enabled by default (in a future release) at the same time as we change the default value of `renderer.outputEncoding` to `sRGBEncoding`, minimizing impact of the output encoding change on the remaining users who haven't yet enabled it.

```js
// current defaults
THREE.ColorManagement.legacyMode = true;
renderer.outputEncoding = LinearEncoding;

// future defaults
THREE.ColorManagement.legacyMode = false;
renderer.outputEncoding = sRGBEncoding;
```

As discussed in https://github.com/mrdoob/three.js/issues/23614, `.outputEncoding` should perhaps also be renamed to `.outputColorSpace` at this time.

RGB and HSL components (e.g. `color.r`) continue to be Linear-sRGB ("Linear"), and can be manipulated exactly as before. I understand this may seem like an inconsistency, but it's exactly how Blender's color picker works as well, and ensures that lerping and keyframe animation continue to execute in Linear-sRGB color space.

```js
// No change for RGB or HSL input.
color.r = color.g = color.b = 0.5;
color.setRGB( 0.5, 0.5, 0.5 );
color.setHSL( 0.0, 1.0, 0.5 );
console.log( color.r ); // → 0.5

// CSS input is sRGB by specification, and converted to Linear-sRGB by this PR.
color.setStyle( 'rgb( 127 127 127 )' );
console.log( color.r ); // → 0.214041140

// Hex input is sRGB by convention, and converted to Linear-sRGB by this PR.
color.setHex( 0x808080 );
console.log( color.r ); // → 0.214041140

// Default input color space can be overridden.
color.setStyle( 'rgb( 127 127 127 )', LinearSRGBColorSpace );
color.setHex( 0x808080, LinearSRGBColorSpace );
console.log( color.r ); // → 0.5
```

**Changes**

- Add `SRGBColorSpace` and `LinearSRGBColorSpace` enums
- Add `colorSpace` argument to THREE.Color getters/setters
    - Hexadecimal and CSS input are considered sRGB by default
    - All other inputs are considered Linear-sRGB by default
    - sRGB input will be converted to Linear-sRGB, as required by the renderer
- Add `ColorManagement` API. Only its `.legacyMode` flag is intended as a public API at this time. Other methods or properties may be added to support wide-gamut workflows in the future.

**Effects**

- Frameworks like A-Frame and R3F already enable color management as a layer on top of three.js, treating user-provided colors as sRGB by default. These changes allow vanilla three.js users to provide colors in the same way.
- Many examples already use `renderer.outputEncoding=sRGB`, and so hard-coded colors (like background and ground colors) are implicitly Linear-sRGB... These will need to be updated (or their appearance will change) with color management enabled.
- Fog and background color are excluded from output encoding. As a result, the two lines below result in different colors, and that doesn't seem ideal:
    - `basicMaterial.color.setHex( ..., sRGBColorSpace )`
    - `background.setHex( ..., sRGBColorSpace )`

**Next steps**

An upcoming PR will add documentation explaining what is meant by "working color space", and how that relates to choices of input and output color spaces. That documentation is meant to be a significantly expanded version of the content at https://www.donmccurdy.com/2020/06/17/color-management-in-threejs/.

/cc @gkjohnson @WestLangley

donmccurdy committed:
ColorManagement: Initial commit

donmccurdy commented on src/math/ColorManagement.js:
```diff
@@ -0,0 +1,124 @@
+			0.0034712266, 0.0380157226, 0.9585130508,
+		],
+	},
+}
```
Display P3 and ACEScg are included to help demonstrate forward-compatibility with wide-gamut workflows, but these could also be removed, or moved to examples/jsm. For example:

```
import { Color, ColorManagement } from 'three';
import { Colors } from 'three/examples/jsm/math/Colors.js';

ColorManagement.addTransform( 'display-p3', 'acescg', Colors.DISPLAYP3_TO_ACESCG );
ColorManagement.addTransform( 'acescg', 'display-p3' Colors.ACESCG_TO_DISPLAYP3 );
```

I don't expect we'll want to support other color spaces in WebGLRenderer. But we may someday want them with WebGPURenderer. And we could use wide-gamut color spaces in SVGRenderer today, with relatively minor changes in the code (for Safari only).

donmccurdy replied:
I've removed Display P3 and ACEScg from the PR for now; they weren't fully usable anyway.

donmccurdy commented on src/math/ColorManagement.js:
```diff
@@ -0,0 +1,124 @@
+
+export const ColorManagement = {
+
+	enabled: false,
```
This flag lets users opt-in to the changes for the first few releases, and opt-out after that. We may or may not want to remove the flag later on.

donmccurdy replied:
Renamed; see thread below.

donmccurdy commented on src/math/ColorManagement.js:
```diff
@@ -0,0 +1,124 @@
+
+	get workingColorSpace () {
+
+		return LinearSRGBColorSpace;
```
To be explained in greater detail in the upcoming documentation PR, but here's a short explanation of "working color space" for now:

_Rendering (and most other operations) must be done using a linear, scene referred color space. The color space assigned for rendering is referred to as the Working Color Space, and is not user-configurable at this time._

donmccurdy replied:
See https://github.com/mrdoob/three.js/pull/23430.

donmccurdy reviewed (commented):
Adding a few additional notes in comments here... I'll go into more detail in an upcoming documentation PR.

WestLangley commented on src/math/ColorManagement.js:
```diff
@@ -0,0 +1,124 @@
+
+		if ( CAT02[ sourceColorSpace ] && CAT02[ sourceColorSpace ][ targetColorSpace ] !== undefined ) {
+
+			_matrix.fromArray( CAT02[ sourceColorSpace ][ targetColorSpace ] ).transpose();
```
Make sure the .transpose() is actually correct -- if you haven't already. It's a function of left-handed vs. right-handed matrix multiplication.

donmccurdy replied:
I did confirm it while implementing. The only example that exercises this CAT02 code path is an experimental wide-gamut SVGRenderer patch, not included in this PR. Mostly I'm including this part now to justify the `to/fromWorkingColorSpace` naming, as opposed to our current `linearToSRGB/SRGBToLinear` method naming.

WestLangley commented on src/math/ColorManagement.js:
```diff
@@ -0,0 +1,124 @@
+		if ( CAT02[ sourceColorSpace ] && CAT02[ sourceColorSpace ][ targetColorSpace ] !== undefined ) {
+
+			_matrix.fromArray( CAT02[ sourceColorSpace ][ targetColorSpace ] ).transpose();
+			_vector.set( color.r, color.g, color.b ).applyMatrix3( _matrix );
```
This method can result in out-of-gamut results. You may have to map the computed values to [0, 1]. There are various ways to do that.

donmccurdy replied:
In a Display P3 → sRGB conversion for example, I think I'd expect this method to clamp the results to [0, 1] rather than doing a continuous remapping. So a clamp should be added here – does that sound ok?

WestLangley replied:
Fine, but remember there may be better approaches than clamping. I would say that topic needs investigation, if you haven't done it already.

donmccurdy replied:
I've removed Display P3 and ACEScg from the PR for now; they weren't fully usable anyway. Let's revisit when we're ready to support these color spaces.

Mugen87 commented:
> Add colorSpace argument to THREE.Color getters/setters, defaulting to sRGB

Why not making it to a property? A color define on app level does usually not change color spaces over time.

Mugen87 commented:
So the color space of the `Color` is the working color space, right? That's a bit different to an earlier outlined approach where the color space would be a property of `Color`. It seems if we are not planning to make the working color space configurable, this solution seems more simple.

WestLangley commented on src/constants.js:
```diff
@@ -145,6 +145,12 @@ export const RGBADepthPacking = 3201;
+export const sRGBColorSpace = 'srgb';
+export const LinearSRGBColorSpace = 'srgb-linear';
+export const DisplayP3ColorSpace = 'display-p3';
+export const LinearACESCGColorSpace = 'acescg';
```
I would name the constant `ACESCGColorSpace`, which is the proper name.

Also, so users do not have to remember when "s" is capitalized and when it is not, I would name it `SRGBColorSpace`.

donmccurdy replied:
`SRGBColorSpace` would be inconsistent with `sRGBEncoding` — is that OK?

Sounds good on `ACESCGColorSpace`.

WestLangley replied:
`sRGBEncoding` should be removed.

donmccurdy replied:
Remaining values in the encoding enum are:

- THREE.LinearEncoding
- THREE.sRGBEncoding
- THREE.BasicDepthPacking
- THREE.RGBADepthPacking

I'm not really familiar with the packing types... would they remain as encodings, or become something else? Setting that aside, removing `texture.encoding` we would have:

- `texture.colorSpace = SRGBColorSpace | LinearSRGBColorSpace`
- `renderer.outputColorSpace = SRGBColorSpace | LinearSRGBColorSpace`

I agree these are clearer and more correct names, but am not sure whether to bundle all of this into one three.js release or not.

WestLangley replied:
Not all textures have a color space, so you would have to allow for that with something like `THREE.RawData` or `THREE.NonColorData` -- or support a `null` value to indicate non-color data for the `.colorSpace` property.

donmccurdy replied:
Great point – I think @gkjohnson had suggested that as well actually. I believe `null` would work well, or maybe `THREE.NoColorSpace`? This should probably be the default, rather than linear, in my opinion.

gkjohnson replied:
> I believe `null` would work well, or maybe `THREE.NoColorSpace`

I far prefer the explicit and more understandable use of an enum name here like `THREE.NoColorSpace` or `THREE.RawData`.

donmccurdy replied:
We might be down to aesthetics at this point. 🙂 I think any of these names are correct enough, and I'm happy to go with any of them. The only downside I see to `RawData` or `NonColorData` is that they don't have the same suffix as other `*ColorSpace` enums, but that's fine with me. Compare the four "in context":

```
texture.colorSpace = null;         // (a)
texture.colorSpace = NonColorData; // (b)
texture.colorSpace = RawData;      // (c)
texture.colorSpace = NoColorSpace; // (d)
```

gkjohnson replied:
I'm not too keen on using "null" myself. (d) `NoColorSpace` would be my preferred option of the above.

gkjohnson commented on src/math/ColorManagement.js:
```diff
@@ -0,0 +1,124 @@
+
+export const ColorManagement = {
+
+	enabled: false,
```
Saying that "Color Management" is disabled by setting `ColorManagement.enabled` equal to `false` doesn't entirely feel correct. In order for color to truly be "unmanaged" here textures need to not be transformed before being passed into the shader which is not how three.js currently works or works after this PR. I've mentioned before that what happens currently feels closer to operating in an sRGB color space but even that isn't a completely accurate depiction of what's currently happening. Three.js' current handling of color and textures is a bit of it's own thing currently. It seems best to refer to it as a "legacy" mode rather than trying to adopt the "Unmanaged Color" term and full deprecate it later. If we really want an option for unmanaged color in the future then more changes to textures will need to be made.

donmccurdy replied:
I think I understand your concerns on this, but I am not sure how best to address it. If we think of this in terms of modes or workflows, there are several, and we may or may not want to support them all. Input colors and textures may come from various color spaces. In each mode ...

1. **Color Managed (NEW)**: Colors and textures are converted to the Linear-sRGB working color space during upload or processing.
2. **Non Color Managed**: Colors and textures are _not_ converted to the Linear-sRGB working color space automatically. This workflow does not exist in three.js today (textures are always converted if marked sRGB), and we may or may not want to add it.
3. **Current / "Legacy"**: Textures are converted to Linear-sRGB during upload; colors are not modified. This is the current behavior, and there are 2-3 likely reasons a user might choose it:
  - a. User is not familiar with color management, uses `outputEncoding=LinearEncoding`, and is effectively using a _gamma_ workflow. This is something I want to discourage and eventually deprecate.
  - b. User is comfortable with color management, uses `outputEncoding=sRGBEncoding`, understands color literals must be provided in Linear-sRGB, and is effectively using a _linear_ workflow.
  - c. User's project does not require lighting, and/or is compositing 2D layers with blending that does not require a linear ('scene referred') working color space. Any color space conversions (especially the additional precision required for internal Linear-sRGB render targets) is just unwanted performance overhead for this user, so they might prefer to do everything in a display-referred color space like sRGB. Sprite-based experiences like [Oat the Goat](http://oatthegoat.co.nz/) might be good examples.

I believe (3c) and (3b) are valid use cases that we should not forget. Perhaps we push these users toward (2) instead of (3), but I don't think we can expect them to be happy adopting (1). So our choice of naming (`ColorManagement.enabled` or otherwise)should be respectful of that.

Ideas:

- Push group (3b) toward option (2), deprecate (3a) and (3b), name the flag `ColorManagement.enabled`
- Don't support (2), name the flag something that invites (3b) and (3c) but discourages (3a) else, like:
    - `Color.enableTransform = false`
    - `ColorManagement.sourceColorSpace = LinearSRGBColorSpace`
    - `ColorManagement.workingColorSpace = sRGBColorSpace`
    - ...

donmccurdy replied:
On further thought, users in group (3b) cannot use workflow (2), because their textures still need to be converted... I guess I'm tempted then to support (1) and (3) only for now, and to decide later whether (2) is necessary and whether (3) can be deprecated.

gkjohnson replied:
My vote is to call the current behavior "Legacy" in some way and then in the long run implement a non color managed mode if we feel it's needed. At the least I just don't think we can use terminology that implies the current three.js behavior is the same as unmanaged color. But I do feel the "Legacy" mode should be deprecated and removed at some point.

I agree that user group "A" is should be discouraged in the long term. And I think user group "C" is really best suited by a true non color managed work flow. Regarding user group "B" --

> On further thought, users in group (3b) cannot use workflow (2), because their textures still need to be converted...

I don't think an unmanaged color workflow would be the easiest path for user group B, anyway. The only thing user group B would have to do to migrate to the new color managed mode is change any place that colors are explicitly set to use an sRGB hex _or_ explicitly pass that the color is in the Linear sRGB color space. It's a migration task but it's not a big one. Realistically I think this is the case that most users are going to fall in to given the current three.js recommendations.

And in terms of converting textures I don't think that's the case as long as the textures are already sRGB. With unmanaged color "texture.colorSpace" and "renderer.outputColorSpace" would be entirely ignored so the lighting calculations be off but the final displayed color would look reasonable as long as sRGB colors are passed into the Color constructor and textures are original stored in sRGB color space.

donmccurdy replied:
> And I think user group "C" is really best suited by a true non color managed work flow...

I'm having second thoughts about (C) now. Consider CSS – it has nothing to do with PBR rendering, but CSS Color Module Level 4 adds support for Display P3 (wide-gamut) colors and actually _does_ need a color managed workflow. How do you feel about ... 

```
ColorManagement.mode = THREE.FullColorManagement;    // (a)
ColorManagement.mode = THREE.LegacyColorManagement;  // (b)
```

... with the caveats that "full" is more aspirational than truth at the moment, and "legacy" should eventually be removed.

gkjohnson replied:
In the long term if Oat the Goat wanted to properly support wider-gamut displays then yes it would need color management enabled. In the short term (or long term if they limit their capability to the narrower sRGB color space) then disabling color management should work.

But overall yeah I agree I think a `ColorManagement.mode` field with a "Legacy" option should work in the short term and should be removed in the long term.

donmccurdy replied:
I think I've settled a boolean `legacyMode` flag, if that sounds OK:

```js
ColorManagement.legacyMode = false; // default 'true'
```

The inverse of that could also work, e.g. `standardMode` defaulting to 'false'.

gkjohnson commented on src/math/Color.js:
```diff
@@ -334,23 +347,33 @@ class Color {
 	}
 
-	getHex() {
+	getHex( colorSpace = sRGBColorSpace ) {
```
We'll need a method to get the r/g/b color components out in an explicit color space. The model exporters, for example, often write the r/g/b components out explicitly in a specific color space. I propose something like `Color.getRGB( target, colorSpace )` where "target" is a basic object with r, g, b fields.

donmccurdy replied:
Sounds good to me!

WestLangley replied:
Instead, I would follow the existing pattern:
```js
color.convertSRGBToLinear();

color.copySRGBToLinear( source_color );
```

The operation is applied to `this`. The inputs are not changed.

donmccurdy replied:
For the use case @gkjohnson is describing (like an reusable exporter, or perhaps retrieving a color to display in a color picker) the advantage of providing a target color space is that it's more robust against different contexts:

1. color management might be disabled – see 3b and 3c above; the value is already sRGB
2. working color space might (in the future) be something other than linear-srgb, like ACEScg

With an API like `color.getRGB( target, sRGBColorSpace )` an exporter can reliably get a color in the expected color space, regardless of other application state.

WestLangley replied:
My point is, it should be `color.convert( source_color, fromColorSpace, toColorSpace );` The args are not changed; "this" is.

Also, how do you convert from one color space to another with out specifying both color spaces?

donmccurdy replied:
^This is how I've been thinking about it as well – if we agree, a `Color` instance is always in the working color space, and only does a conversion to other color spaces when serialized. The main flies in that ointment are `scene.fog` and `scene.background`, which should _not_ be given in the working color space at the moment. But those would change if we do the output transforms in a post-processing pass rather than the shader.

gkjohnson reviewed (commented):
Really glad to see how this has come along! A few thoughts other thoughts:

1. In docs it should be very explicitly noted that the `Color.r/g/b` members are _not_ in a reliable color space. Not only because of this legacy mode support but in the future if other wider gamut color spaces are supported then it will depend on the current working color space.

2. Should functions like `Color.convertLinearToSRGB` log a deprecation warning when the legacy color path is enabled since they're inconsistent with the new color storage model?

3. I assume we'll only be changing `WebGLRenderer.outputEncoding` to `sRGBEncoding` by default once this legacy mode is disabled by default?

Overall I'm happy with what this has evolved in to. The API surface is consistent (aside from the Color.r/g/b fields) which was my original concern when improving the color spaces for the project.

donmccurdy commented:
> Would you please explain why the Color class needs a different approach for handling color spaces than Texture? Textures allows to define the color space per texture. Why can't we do this for Color? TBH, I still like the Color.encoding approach^^.

Let me come back to this question once I've opened the documentation PR, I think my answer will not be very clear without that additional context.

WestLangley commented:
>I assume we'll only be changing WebGLRenderer.outputEncoding to sRGBEncoding by default

@gkjohnson I would rename the property:

    renderer.outputEncoding -> renderer.outputColorSpace

donmccurdy commented:
Update —

- https://github.com/mrdoob/three.js/pull/23430

@Mugen87 to your question earlier, I believe that almost anything a user will 'do' with a Color instance should done in the working color space (Linear-sRGB, as defined in link above), and that's _not_ the color space users will find in a color picker or their CSS stylesheets, which are usually sRGB.

If we define `color.encoding` or `color.colorSpace` and most colors get marked sRGB — since that is what users initialize them with! — there's very little you can safely do with the RGB components of that Color instance. Instead I think it's best to specify the color space of the input when constructing a color, and immediately convert to Linear-sRGB at that point.

donmccurdy commented:
As there are a few steps here, my suggested grouping of PRs would be:

**r138**

- #23392
- #23430
- color management _disabled_ by default, begin opt-in testing

**r139**

- rename `texture.encoding` → `texture.colorSpace`
- rename `renderer.outputEncoding` → `renderer.outputColorSpace`
- change default renderer output to sRGB
- color management _enabled_ by default

Wouldn't need to be r138 and r139 specifically of course, but basically I think these changes should be spread over two releases, with the breaking changes grouped together.

WestLangley commented:
The `Color` class is NOT associated with a color space.

As I have said, the API to follow is:
```js
color.convertSRGBToLinearSRGB();

color.copySRGBToLinearSRGB( source_color );
```
You need a source color space and a destination color space.

There are not that many variations, so hardwiring a separate method for each pair seems OK to me.

//

Alternatively, you can provide:
```js
color.convert( fromColorSpace, toColorSpace );
color.copy( source_color, fromColorSpace, toColorSpace );
```

donmccurdy commented:
> The Color class is NOT associated with a color space.

This feels like a decision we need to make, not a fact of the universe — I believe there are advantages to associating the RGB components of the Color class with the working color space (currently Linear-sRGB), as described above. Could you help me understand what advantages you see in the API you are suggesting, or the disadvantages in this PR? Am I correct in understanding that you'd like to have these convert/copy methods and _not_ have these parameters on setters/getters?

If we only offer methods that convert a color from A to B,  it seems like we are still expecting all Color instances to be provided on materials in Linear-sRGB, but we are making it harder for users to do that — because each needs to be explicitly converted from sRGB. I would like this syntax to 'just work', without needing to be told the hex color is sRGB:

```js
const material = new MeshStandardMaterial( { color: 0x502020 } );
```

WestLangley commented:
>The Color class is NOT associated with a color space.
>>This feels like a decision we need to make, not a foregone conclusion

The statement is true. But, yes, the implication of my statement is it should stay that way. Just my humble opinion. :-)

>Am I correct in understanding that you'd like to have these convert/copy methods and not have these parameters on setters/getters?

Yes.

> I would like this syntax to 'just work', without needing to be told the hex color is sRGB:

IMO, your approach results in too much magic behind the scenes. That is just my view. I know your concept is proposed with good-intention.

//

Good news, thanks to recent work by @gkjohnson:

1. Loaders now work correctly and convert colors to scene-referred linear. Users just need to set the proper output color space for rendering. That is a matter of copying the three.js pattern.

2. If a user is using a color picker, then the user needs to know something about color space conversion. Again, it is a matter of copying the three.js patterns in the examples -- or adding appropriate examples if they do not exist.

gkjohnson commented:
I don't feel that enforcing color storage in a common color space is all that magic. If we're going to hope to ensure that a color is in our working color space by the type they get to a shader then we need to have some reliable way of knowing what space a color instance is in.

Another option for enabling that is to store colorSpace on the Color instance itself but that requires converting the color to a common color space whenever an operation is done (lerp, add, etc) and converting before uploading to the shader. But that feels unnecessarily complicated.

donmccurdy commented:
> Another option for enabling that is to store colorSpace on the Color instance itself ... But that feels unnecessarily complicated.

Yeah, unfortunately I think this way would get too complicated.

> IMO, your approach results in too much magic behind the scenes ...

> I don't feel that enforcing color storage in a common color space is all that magic ...

There'd be magic in saying, "all three.js methods accepting Color instances assume the color's RGB components are Linear-sRGB". It's not totally true yet — see `scene.fog` and `scene.background` — but mostly true and we could fix the rest if we want to. I don't think we're disagreeing about that.

The harder question is whether using sRGB for getters/setters (hex, CSS, HSL) and Linear-sRGB for floating-point RGB components is too much magic. Personally I think it's the right choice — A-Frame and R3F read/write sRGB hex colors, and internally convert to Linear-sRGB for THREE.Color instances. As far as I can tell Filament [makes the color space a _required_ argument](https://google.github.io/filament/webgl/tutorial_redball.html) when setting a color on a material.

Either way, it's a big change — see implications under _Working with THREE.Color Instances_ in #23430. I'm hoping more people will weigh in on that part of the API.

WestLangley commented:
>There'd be no magic in saying, "all three.js methods and properties using Color instances assume the color's RGB components are Linear-sRGB".

There will be confusion, IMO.

Vertex colors in glTF are linear multipliers, while the base color in glTF is assumed to be encoded with the sRGB transfer function.

You are trying to enforce a particular color space with `THREE.Color`. `THREE.Color` is not associated with a color space.

donmccurdy commented:
> Vertex colors in glTF are linear multipliers...

The glTF specification's phrasing of "linear multipliers" is something I'd prefer to leave out of three.js. It's clearer to say that both three.js vertex colors, `material.color`, `light.color`, etc. are expected to be given in Linear-sRGB space. The same is true of almost any role colors may fill in three.js. If we want to say that a loose/unused THREE.Color does not have a color space that's fine, but a color used by a material or a light must be Linear-sRGB.

> If a user is using a color picker, then the user needs to know something about color space conversion...

The current state of three.js is that every time the user provides a color from virtually any other software, they must immediately do an explicit sRGB → Linear-sRGB conversion. The same is true for our own examples using `dat.gui` color pickers. It's good for users to know something about color space conversion, but requiring boilerplate conversion on every color input is not a good status quo, and not (IMO) standard practice in a graphics library. The objective of this PR is to ensure that a correct color workflow (with sRGB → Linear-sRGB conversion) is also the easiest color workflow. If this requires adding a stated assumption that certain THREE.Color instances contain RGB components in Linear-sRGB, I think we should do that.

WestLangley commented:
@donmccurdy OK. I do not want to be a blocker. Your approach deserves a try. If others are happy with it, that is great. :-)

donmccurdy commented:
@WestLangley thank you for being open on this. I don't want to force this through if it's divisive. Perhaps let's leave the PR up for a while and see if we can gather more feedback.

The [`<color>` datatype in CSS](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value) is another interesting example, as some syntax is inherently sRGB, and other syntax (with Color Module Level 4) has other defaults or is configurable.

Pulling in a few people from #11337 — @bhouston and @zomboh do either of you have preferences here, or examples from experience in other tools?

bhouston commented:
This makes sense and solves an issue I ran into with Three.js back in 2016: https://github.com/mrdoob/three.js/issues/8827

I think that in the future, we will likely start to support more output encodings rather than just sRGB, specifically HDR outputs, but that requires browser support first, which is coming: https://twitter.com/BenHouston3D/status/1481979159077339143

To speed up the transition to ColorManagement = true, we should likely have a warning for when it is off.  I think that we should aim to actually just remove the idea of ColorManagement.enabled in the future because supporting multiple paths is annoying and this one is clearly right.  If people want the old behaviour just use the ColorSpace specificers when setting/getting Color values.

bhouston commented:
```
// current defaults
THREE.ColorManagement.enabled = false;
renderer.outputEncoding = LinearEncoding;

// future defaults
THREE.ColorManagement.enabled = true;
renderer.outputEncoding = sRGBEncoding;
```

The above confuses me.  I am so confused that renderer.outputEncoding is linear by default.  I thought it was sRGB for the longest time?  outputEncoding linear is simply wrong and will lead to screwed up color output.  I guess over here at @Threekit, we would just always set outputEncoding = sRGB and forgot that Three.js was using the wrong default?

bhouston commented:
Ah, I understand the mistake.  "Encodings" and "ColorSpaces" are being mixed together.

Originally we thought of these options:
* ColorSpace (Linear, sRGB, AdobeRGB, Gamma20, Gamma22)
* Encoding (None, RGBE, RGBM, RGBD, LogLUV, CMYK) parameter.

Per: https://github.com/mrdoob/three.js/pull/11076#issuecomment-293570510

But I think that the above dichotomy isn't completely accurate.  RGBE, RGBM, LogLUV are the same linear HDR colorspace but represented in 4 bytes (which some clipping - maybe that makes them different colorspaces?.)  Where as CMYK is an LDR color space without Alpha.

I think we need better clarify on ColorSpace and Encoding usage.

donmccurdy commented:
There are fewer encodings left now (see https://github.com/mrdoob/three.js/pull/23392#discussion_r798880542), so I agree with @WestLangley's previous suggestion that sRGBEncoding should be removed and replaced with `*ColorSpace` alternatives.

> outputEncoding linear is simply wrong and will lead to screwed up color output...

I think we've all agreed for a while that the default should be changed to sRGB – but I'm at least partly responsible for why that hasn't happened yet, as I've been pushing to wait and change the output defaults and the input defaults (for hex and CSS colors) at the same time: https://github.com/mrdoob/three.js/pull/22275#issuecomment-900000676.

DennisSmolek commented:
> but that requires browser support first, which is coming: https://twitter.com/BenHouston3D/status/1481979159077339143

From what I can understand it's already here to some extent: https://developer.chrome.com/blog/new-in-chrome-94/#canvas-colorspace
The canvas can now be set to at least P3 and I've read it said it can be set to Rec. 2020 too and only depends on if the display/browser/os can support it.
This Webkit page is pretty great too and has detectors that work in real time. My iMac supports P3 but as I drag to other monitors it adjust and a warning appears.

DennisSmolek commented:
After being sparked by Don's tween earlier I've (yet again) spent far too long going down the color rabbit hole..

In the UI/Design side we've had color pains in the web resulting in some amazing efforts to not only educate. but also provide solutions.

Even with all that I still struggle to really get what's going on or know what the "right" action is to take.. Converting in/blending/how it'll react with lights etc. On top of all the other things a person faces when trying something new. 

### So in regards to if you should convert with `setHex()` etc:
I appreciate that Three is adding more documentation to help (me) and others to get what's happening, but I also think it a GREAT idea to just take care of some of it by default.

I think it smart to lower the barrier to entry where you can, but not obscure or hide what's going on, rather I'd suggest offering the **Progressive Enhancement** of diving into the nitty gritty of it all when the user is ready/curious/problem solving.

And for those who've already got the experience & skillset they can use the overrides to dial their workflow to their EXACTING specifications..

donmccurdy commented:
> From what I can understand [wide gamut canvas is] already here to some extent...

Unfortunately limited to Canvas 2D contexts – no support for WebGL yet. But eventually it will happen, in WebGPU if not in WebGL.

> I appreciate that Three is adding more documentation to help (me) and others to get what's happening, but I also think it a GREAT idea to just take care of some of it by default. ... I think it smart to lower the barrier to entry where you can, but not obscure or hide what's going on, rather I'd suggest offering the Progressive Enhancement ...

Thanks @DennisSmolek! Improving defaults (for both correctness and ease of use) is a goal of this PR, along with the related changes to `renderer.outputEncoding`. The aim of progressive enhancement is a helpful one, e.g. any options related to wide-gamut rendering should be fully "opt-in". But, similar to Blender's requirement of marking color and non-color textures, I think certain concepts can't fully be hidden.

donmccurdy commented:
After starting a [thread on Twitter](https://twitter.com/donrmccurdy/status/1490452069043654662?s=20&t=wm9ko9_tgnDgd9LQX845bA) (thanks for the comments and questions @gkjohnson and @bhouston), there's an interesting question on specifically _which_ inputs should be considered sRGB...

| syntax  | blender | three.js |
|---------|---------|----------|
| RGB     | linear  | ?        |
| HSL/HSV | linear  | ?        |
| Hex     | sRGB    | sRGB     |
| CSS     | —       | sRGB     |

I think it's important that the `.r` `.g` `.b` components of a Color be Linear-sRGB, at least when the Color is used for `material.color` or similar roles. But either default could be rational for methods like setRGB and setHSV. I'm inclined to follow Blender, but could certainly be talked out of that. For example:

- `setHex( hex, colorSpace = sRGBColorSpace )`
- `setStyle( style, colorSpace = sRGBColorSpace )`
- `setRGB( r, g, b, colorSpace = LinearSRGBColorSpace )`
- `setHSL( h, s, l, colorSpace = LinearSRGBColorSpace )`

This still leaves the option of overriding the default, e.g. for an exporter that needs RGB components in sRGB color space.

gkjohnson commented:
Either solution for `setRGB` or `setHSL` will be perfectly usable but if I were going to pick between the two I think I'd pick sRGB as the default for the following reasons:

- CSS colors on the web defaults to sRGB even when using ratios (`rgb( 50%, 50%, 50% )`) or hsl (`hsl( 0, 0, 50% )`) and with three.js being a web tool it makes sense to align with the rest of the web ecosystem.
- Using `SRGBColorSpace` makes the default consistent with the rest of the Color API.
- If we're already talking about other wide-gamut color spaces like `ACESCGColorSpace` in the future will the default of `LinearSRGBColorSpace` still make sense in those cases?

That aside I don't want to delay progress on this over figuring out defaults. Just seemed worth noting!

donmccurdy commented:
Did some digging here, and there seems to be a pretty good precedent for treating hexadecimal values as sRGB, _however_, the convention is not so clear for other syntax. If we decide to consider hexadecimal and CSS inputs sRGB, and leave everything else alone, we'll be consistent with the tools below:

## Consistent

✅ Blender:
- Hex = sRGB
- Other = Scene Linear
- Source: https://docs.blender.org/manual/en/latest/interface/controls/templates/color_picker.html — "In Blender, the RGB and HSV/HSL values are in Scene Linear color space, and are therefore not Gamma corrected. On the contrary, Hex are automatically Gamma corrected for the sRGB Color Space."

✅ Maya
- Picker color space configurable, defaults to Scene Linear ("Rendering Space"). Alternative ("Display Space") available. From screenshots I see only RGB components, no hexadecimal input.
- Source: https://knowledge.autodesk.com/support/maya/learn-explore/caas/CloudHelp/cloudhelp/2018/ENU/Maya-Rendering/files/GUID-EBC11F85-9CAB-4D33-ABF4-836AD47D5C76-htm.html

✅ Unreal Engine:
- Hex = sRGB or Linear (two inputs in color picker)
- Other = Scene Linear
- Source: https://docs.unrealengine.com/4.27/en-US/Basics/UI/ColorPicker/

## Mostly sRGB

⚠️ CSS:
- Everything is sRGB unless otherwise specified.

## Unknown

⚠️ Unity
- Color picker: Unknown https://docs.unity3d.com/Manual/EditingValueProperties.html#color-values
- Color class: Everything sRGB, class has .linear property to access Linear-sRGB values: https://docs.unity3d.com/ScriptReference/Color-linear.html

⚠️ Substance Painter
- Unable to find a clear answer.

⚠️ Houdini
- Unable to find a clear answer.

🛑 3DS Max
- Not color managed

donmccurdy commented:
> If we're already talking about other wide-gamut color spaces like ACESCGColorSpace in the future will the default of LinearSRGBColorSpace still make sense in those cases?

Based on Blender, Maya, and Unreal Engine I would say that if we allow the working color space to be changed in the future, we should also change the default for inputs other than hex and CSS. But that shouldn't be a breaking change with an API like:

```js
class Color {

  setRGB( r, g, b, colorSpace = ColorManagement.workingColorSpace ) {

    // ...

  }

}
```

donmccurdy commented:
Updates —

- Removed Display P3 and ACEScg code from this PR; more work is needed to make them useful.
- Limited the scope of this change to only assume **hexadecimal** and **CSS** inputs are sRGB, the rest are assumed to be Linear-sRGB. This matches up well with Blender, Maya, Unreal Engine, and the CSS Color Module Level 4 specification — which seems like as good an outcome as could be hoped for.

***

Coming back to @WestLangley's concerns that `THREE.Color` should not be associated with a color space. Would this change seem better if the association were not global, and a `Color` instance could be associated explicitly with other color spaces in appropriate contexts? We could explicitly mark `scene.fog` and `scene.background` (or other examples) as sRGB in that case. This brings us back to @Mugen87's comments on `color.colorSpace`. As long as the default for most colors is Linear-sRGB (or the working color space) and _not_ sRGB, I think that could work. Examples:

```js
// Defaults.
material.color.colorSpace = LinearSRGBColorSpace;
scene.fog.color.colorSpace = SRGBColorSpace;

// Automatic conversion to from sRGB to Linear-sRGB.
material.color.setHex( 0x880000 );

// No conversion — input and color are both sRGB.
scene.fog.color.setHex( 0x880000 );
```

**Note:** I'm not arguing that fog _should_ be sRGB, but that's currently how it functions in the rendering pipeline. Perhaps there are better examples of when `color.colorSpace` should have values other than Linear-sRGB.

Mugen87 commented:
> Would this change seem better if the association were not global, and a Color instance could be associated explicitly with other color spaces in appropriate contexts?

I favor this model since it is more flexible.

If we decide to use `Color.colorSpace`, we should probably rename `Texture.encoding` so both properties are aligned. It would be great if users needs to understand just a single color/texture property to implement a proper color space workflow.

zomboh commented:
@donmccurdy sorry it took me so long to give feedback! Not much to say though, there's been quite a bit of progress - I think you're making good calls, the alignment with Blender etc makes perfect sense and it's the way I would expect three to interpret color if I was a newcomer, so... good defaults, many thanks for working on this!

+1 as well to the Color instance model to associate color spaces depending on context.

donmccurdy committed:
Color: revert to original CSS whitespace.

donmccurdy commented:
**NOTICE:** I've just rewritten the first few paragraphs of this PR's description, to reflect the feedback and changes to the PR. Specifically, calling out the fact that _only_ hexadecimal and CSS input are considered sRGB and converted.

***

@Mugen87 after some more discussion and mulling over, I think I'd like to delay the question of adding `color.colorSpace` until some of the other changes of https://github.com/mrdoob/three.js/issues/23614 are in place. I'm not sure about parts of it yet. This PR will function the same way with or without that change.

***

In my opinion, this PR brings THREE.Color getter/setter APIs to where they ought to be in a linear workflow, while accommodating conventions our users expect from experience with CSS and Blender. Certain naming choices (e.g. `NonColorData` and `ColorManagement.legacyMode`) could perhaps use a bit more fine-tuning, but functionally I believe this is the right solution. @mrdoob — this should be ready for your review and feedback. Thanks!

mrdoob commented:
Does @WestLangley approve? 😅

mrdoob commented on src/constants.js:
```diff
@@ -145,6 +145,11 @@ export const RGBADepthPacking = 3201;
 export const ObjectSpaceNormalMap = 1;
 
+// Color space string identifiers, matching CSS Color Module Level 4 and WebGPU names where available.
+export const NonColorData = '';
```
We prefer `NonColorData` better than `NoColorSpace`?

donmccurdy replied:
I've been staring at these terms too long, they all blur together now. I'm completely happy with either. NoColorSpace has a nice parallelism with NoToneMapping... Do you have a preference?

mrdoob replied:
I vote for `NoColorSpace` yeah.

donmccurdy replied:
Sounds good — done: [68407d7](https://github.com/mrdoob/three.js/pull/23392/commits/68407d7c0a89eefdd2a187cccabdf0ed0cde892b)

gkjohnson commented:
Regarding adding `Color.colorSpace` I think it's worth noting that I don't think it simplifies any other part of the API and just adds overhead and complexity imo.

- A global color space must still be set for shaders and color operations but now colors must be converted to and from the working color space when performing math operations and updating shader uniforms.
- The constructor, setter, and getter APIs must still all have an argument to specify a target color space to retrieve and set color data in.

I guess I'm just not sure I see the benefit.

Another thing that I don't think we've really thought about, though I suppose it's longer term, is how vertex colors are handled. It's easy when we say Linear sRGB is the working color space and they're stored as such but what happens when P3 or ACES are the working color space? If vertex colors are going to be interpolated correctly in a shader or with animations then they must have either a color space associated with them or be implicitly converted to the correct color space. I think a solution for that can wait but I thought I'd mention it in the interest of being thorough.

donmccurdy commented:
> I don't think [`color.colorSpace`] simplifies any other part of the API...

A few quick thoughts. If we added such a property, its default should certainly be the working color space, Linear-sRGB (not sRGB). Automatic conversions should be generally avoided — if an operation is valid on the given color space it should run, otherwise the operation should probably warn or fail — to avoid a whole mess of complexity. For instances managed by three.js (like `material.color`) it would _not_ be valid to change the color space to anything other than the working color space as used in the shader.

I do think use good cases exist — e.g. working with LightProbes or LUTs that don't need to be constrained to the sRGB gamut, perhaps? But I don't feel sure about those use cases yet, and I don't think easier Linear-sRGB ↔️ sRGB conversion is the right application for it.

> Another thing that I don't think we've really thought about, though I suppose it's longer term, is how vertex colors are handled...

I think it's fair to say that any wide-gamut working color space will require considerable care; nothing in any 3D model format will work 'out of the box'. 😢 But ultimately I think we'd expect vertex colors to end up in the working color space, and it's straightforward to pre-process glTF files for that purpose.

donmccurdy committed:
Constants: NonColorData → NoColorSpace

mrdoob commented:
Thanks!

gkjohnson commented:
> A few quick thoughts. If we added such a property, its default should certainly be the working color space, Linear-sRGB (not sRGB). **Automatic conversions should be generally avoided** — if an operation is valid on the given color space it should run, otherwise the operation should probably warn or fail — to avoid a whole mess of complexity.

I guess to that end I'm not seeing what the real benefit is to the flag. It seems to make things more confusing. I just imagine users wondering why they can't interpolate between two colors. And if two colors must be in the same color space to be operated on together then we're not adhering very strictly to our concept of a "working color space" across three.js, then.

> I think it's fair to say that any wide-gamut working color space will require considerable care; nothing in any 3D model format will work 'out of the box'. 😢 But ultimately we'd expect vertex colors to end up in the working color space, and it's straightforward to pre-process glTF files for that purpose.

I want to believe there's a better solution here but I acknowledge that it's a future problem. Something like a `ColorBufferAttribute` could be added that implicitly converts to the working color space, for example, or a color space could be included alongside the `BufferAttribute` similar to textures so vertex colors can be converted in shaders since byte arrays are not well suited to linear color data.

gkjohnson commented:
Perhaps a line item should be added to https://github.com/mrdoob/three.js/issues/23614 to think about handling of Color BufferAttributes?

donmccurdy commented:
I would say uint8 is not ideal for Linear-sRGB vertex color data, but uint16 and float32 are fine. glTF is (as far as I know?) the only format we have that supports anything more compact than float32, and it specifies Linear-sRGB vertex colors. So I feel like sRGB vertex colors in uint8 buffers don't seem to justify a new BufferAttribute API and shader permutations here. Perhaps a new issue if there's more to consider on that side? I'm not ready to think about vertex colors in P3 or ACEScg quite yet I'm afraid. 😓

gkjohnson commented:
> I would say uint8 is not ideal for Linear-sRGB vertex color data, but uint16 and float32 are fine. glTF is (as far as I know?) the only format we have that supports anything more compact than float32, and it specifies Linear-sRGB vertex colors.

I agree -- this isn't just about file formats, though. It's about enabling users to consistently use, operate on, and render color in three.js with or without existing model formats.

> To the extend that we can avoid conversions and store data in working color space, that is a very valuable thing. Perhaps a new issue if there's more to consider on that side? I'm not ready to think about vertex colors in P3 or ACEScg quite yet I'm afraid. 😓

I just suggest adding it to https://github.com/mrdoob/three.js/issues/23614 in order to keep everything consolidated understanding that it's not an immediate priority. Perhaps as the last line item or as a (2.0+) category bullet.

donmccurdy commented:
Sounds good! We'll have to explore more, but added this under 2.x in https://github.com/mrdoob/three.js/issues/23614 for now.
